kornia/geometry/depth.py

```diff
@@ -0,0 +1,170 @@
+
+    # compute the pointcloud spatial gradients
+    gradients: torch.Tensor = spatial_gradient(xyz)  # Bx3x2xHxW
+
+    # compute normals
+    a, b = gradients[:, :, 0], gradients[:, :, 1]  # Bx3xHxW
+
+    normals: torch.Tensor = torch.cross(a, b, dim=1)  # Bx3xHxW
+    return F.normalize(normals, dim=1, p=2)
+
+
+def warp_frame_depth(
+        image_src: torch.Tensor,
+        depth_dst: torch.Tensor,
+        src_trans_dst: torch.Tensor,
+        camera_matrix: torch.Tensor) -> torch.Tensor:
+    """Warp a tensor from a source to destination frame by the depth in the destination.
+
+    Compute 3d points from the depth, transform them using given transformation, then project the point cloud to an
+    image plane.
+
+    Args:
+        image_src (torch.Tensor): image tensor in the source frame with shape (BxDxHxW).
+        depth_dst (torch.Tensor): depth tensor in the destination frame with shape (Bx1xHxW).
+        src_trans_dst (torch.Tensor): transformation matrix from destination to source with shape (Bx4x4).
+        camera_matrix (torch.Tensor): tensor containing the camera intrinsics with shape (Bx3x3).
+
+    Return:
+        torch.Tensor: the warped tensor in the source frame with shape (Bx3xHxW).
+
+    """
+    if not isinstance(image_src, torch.Tensor):
+        raise TypeError(f"Input image_src type is not a torch.Tensor. Got {type(image_src)}.")
+
+    if not len(image_src.shape) == 4:
+        raise ValueError(f"Input image_src musth have a shape (B, D, H, W). Got: {image_src.shape}")
+
+    if not isinstance(depth_dst, torch.Tensor):
+        raise TypeError(f"Input depht_dst type is not a torch.Tensor. Got {type(depth_dst)}.")
+
+    if not len(depth_dst.shape) == 4 and depth_dst.shape[-3] == 1:
+        raise ValueError(f"Input depth_dst musth have a shape (B, 1, H, W). Got: {depth_dst.shape}")
+
+    if not isinstance(src_trans_dst, torch.Tensor):
+        raise TypeError(f"Input src_trans_dst type is not a torch.Tensor. "
+                        f"Got {type(src_trans_dst)}.")
+
+    if not len(src_trans_dst.shape) == 3 and src_trans_dst.shape[-2:] == (3, 3):
+        raise ValueError(f"Input src_trans_dst must have a shape (B, 3, 3). "
+                         f"Got: {src_trans_dst.shape}.")
+
+    if not isinstance(camera_matrix, torch.Tensor):
+        raise TypeError(f"Input camera_matrix type is not a torch.Tensor. "
+                        f"Got {type(camera_matrix)}.")
+
+    if not len(camera_matrix.shape) == 3 and camera_matrix.shape[-2:] == (3, 3):
+        raise ValueError(f"Input camera_matrix must have a shape (B, 3, 3). "
+                         f"Got: {camera_matrix.shape}.")
+    # unproject source points to camera frame
+    points_3d_dst: torch.Tensor = depth_to_3d(depth_dst, camera_matrix)  # Bx3xHxW
+
+    # transform points from source to destionation
+    points_3d_dst = points_3d_dst.permute(0, 2, 3, 1)  # BxHxWx3
+
+    # hack to match sizes
+    batch_size: int = image_src.shape[0]
```

Comment:
I though that pytorch convention is to use .size(0) over .shape[0].

Reply:
not sure if it's a convention, I use .shape to make it more numpy like. We can decide on that.